### PR TITLE
Query API features added - MUV extraction, VTS MostRecent, FillerStra…

### PR DESCRIPTION
**Pull request by MichaelGeraghty**
…tegy

### CODE REVIEW

File: Artesian/Artesian.SDK.Tests/ActualTimeSerieQueries.cs [hunk 1]
```diff
@@ -482,5 +482,118 @@ public void ActWithHeaders_FilterId()
...
+        {
+            using (var httpTest = new HttpTest())
+            {
+                var qs = new QueryService(_cfg);
+
+                var act = qs.CreateActual()
+                        .ForMarketData(new int[] { 100000001 })
+                        .InGranularity(Granularity.Day)
+                        .InAbsoluteDateRange(new LocalDate(2018, 6, 22), new LocalDate(2018, 7, 23))
+                        .WithFillCustom(123)
```
Comment (AndreaCuneo) [inline Artesian/Artesian.SDK.Tests/ActualTimeSerieQueries.cs]: rename to WithFillCustomValue

File: Artesian/Artesian.SDK.Tests/ActualTimeSerieQueries.cs [hunk 2]
```diff
@@ -482,5 +482,118 @@ public void ActWithHeaders_FilterId()
...
+
+                var act = qs.CreateActual()
+                        .ForMarketData(new int[] { 100000001 })
+                        .InGranularity(Granularity.Day)
+                        .InAbsoluteDateRange(new LocalDate(2018, 6, 22), new LocalDate(2018, 7, 23))
+                        .ExecuteAsync().Result;
+
+                httpTest.ShouldHaveCalled($"{_cfg.BaseAddress}query/v1.0/ts/Day/2018-06-22/2018-07-23"
+                 .SetQueryParam("id", 100000001)
+                 .SetQueryParam("fillerK", FillerKind.Null))
```
Comment (AndreaCuneo) [inline Artesian/Artesian.SDK.Tests/ActualTimeSerieQueries.cs]: This is not expected. 
Either fillerK is Default or is omitted (as is now).
Which strategy is the Default is implemented server-side, not client side.

File: Artesian/Artesian.SDK.Tests/MarketAssessmentQueries.cs [hunk 3]
```diff
@@ -421,6 +421,133 @@ public void MasWithHeaders_FilterId()
...
+                var mas = qs.CreateMarketAssessment()
+                       .ForMarketData(new int[] { 100000001 })
+                       .ForProducts(new string[] { "M+1", "GY+1" })
+                       .InAbsoluteDateRange(new LocalDate(2018, 1, 1), new LocalDate(2018, 1, 10))
+                       .ExecuteAsync().Result;
+
+                httpTest.ShouldHaveCalled($"{_cfg.BaseAddress}query/v1.0/mas/2018-01-01/2018-01-10"
+                    .SetQueryParam("id", 100000001)
+                    .SetQueryParam("p", new string[] { "M+1", "GY+1" })
+                    .SetQueryParam("fillerK", FillerKind.LatestValidValue)
```
Comment (AndreaCuneo) [inline Artesian/Artesian.SDK.Tests/MarketAssessmentQueries.cs]: same as per CreateActual, the Default is implemented server-side, not client side.

File: Artesian/Artesian.SDK.Tests/MarketAssessmentQueries.cs [hunk 4]
```diff
@@ -421,6 +421,133 @@ public void MasWithHeaders_FilterId()
...
+        {
+            using (var httpTest = new HttpTest())
+            {
+                var qs = new QueryService(_cfg);
+
+                var mas = qs.CreateMarketAssessment()
+                       .ForMarketData(new int[] { 100000001 })
+                       .ForProducts(new string[] { "M+1", "GY+1" })
+                       .InAbsoluteDateRange(new LocalDate(2018, 1, 1), new LocalDate(2018, 1, 10))
+                       .WithFillLatestValue(Period.FromDays(14))
```
Comment (AndreaCuneo) [inline Artesian/Artesian.SDK.Tests/MarketAssessmentQueries.cs]: test a period different from the "default" to avoid mis testing.

File: Artesian/Artesian.SDK.Tests/MarketAssessmentQueries.cs [hunk 5]
```diff
@@ -421,6 +421,133 @@ public void MasWithHeaders_FilterId()
...
+        {
+            using (var httpTest = new HttpTest())
+            {
+                var qs = new QueryService(_cfg);
+
+                var mas = qs.CreateMarketAssessment()
+                       .ForMarketData(new int[] { 100000001 })
+                       .ForProducts(new string[] { "M+1", "GY+1" })
+                       .InAbsoluteDateRange(new LocalDate(2018, 1, 1), new LocalDate(2018, 1, 10))
+                       .WithFillCustom(123,456,789,321,654,987,213,435)
```
Comment (AndreaCuneo) [inline Artesian/Artesian.SDK.Tests/MarketAssessmentQueries.cs]: If the SDK already include a MarketAssessmentValue of some sort (it should) please use that one.
Something like
`
.WithFillCustom(new MarketAssessmentValue { Settlement = 12 })
`

File: Artesian/Artesian.SDK.Tests/MarketAssessmentQueries.cs [hunk 6]
```diff
@@ -421,6 +421,133 @@ public void MasWithHeaders_FilterId()
...
+                    .SetQueryParam("fillerDVvp", 987)
+                    .SetQueryParam("fillerDVvg", 213)
+                    .SetQueryParam("fillerDVvt", 435)
+                    )
+                    .WithVerb(HttpMethod.Get)
+                    .Times(1);
+            }
+        }
+        #endregion
+
```
Comment (AndreaCuneo) [inline Artesian/Artesian.SDK.Tests/MarketAssessmentQueries.cs]: Please add other tests to test that each Value can be NULL or not passed. Means that in the query params I expect only the not null values.

File: Artesian/Artesian.SDK/Service/Query/ActualQuery.cs [hunk 7]
```diff
@@ -186,6 +232,27 @@ protected sealed override void _validateQuery()
 
             if (_queryParamaters.Granularity == null)
                 throw new ApplicationException("Extraction granularity must be provided. Use .InGranularity() argument takes a granularity type");
+
+            if (_queryParamaters.FillerKind == FillerKind.Default)
+            {
+                _queryParamaters.FillerKind = FillerKind.Null;
```
Comment (AndreaCuneo) [inline Artesian/Artesian.SDK/Service/Query/ActualQuery.cs]: nope. read other comments

File: Artesian/Artesian.SDK/Service/Query/IMasQuery.cs [hunk 8]
```diff
@@ -1,10 +1,16 @@
...
 // Licensed under the MIT License. See LICENSE in the project root for
 // license information. 
+using NodaTime;
+
 namespace Artesian.SDK.Service
 {
     interface IMasQuery<T>: IQuery<T>
     {
         T ForProducts(params string[] products);
+        T WithFillNull();
```
Comment (AndreaCuneo) [inline Artesian/Artesian.SDK/Service/Query/IMasQuery.cs]: All these methods but WithFillCustomValue should be in the IQuery<T> interface.

Only WithFillCustomValue should be specialized in IActualQuery and IMasQuery

File: Artesian/Artesian.SDK/Service/Query/IVersionedQuery.cs [hunk 9]
```diff
@@ -10,7 +10,11 @@ interface IVersionedQuery<T> : IQuery<T>
     {
         T InGranularity(Granularity granularity);
         T ForLastNVersions(int lastN);
-        T ForMUV();
+        T ForMUV(LocalDateTime? versionLimit);
```
Comment (AndreaCuneo) [inline Artesian/Artesian.SDK/Service/Query/IVersionedQuery.cs]: This is a breaking change.

Any code with ForMUV() will break as the method is no longer available.
Leave ForMUV() with no parameters and make the versionLimit not nullable.
Or make versionLimit? default to NULL.

File: Artesian/Artesian.SDK/Service/Query/QueryParamaters/ActualQueryParamaters.cs [hunk 10]
```diff
@@ -29,19 +30,27 @@ public ActualQueryParamaters()
...
         public ActualQueryParamaters(
             IEnumerable<int> ids, 
             ExtractionRangeSelectionConfig extractionRangeSelectionConfig, 
             ExtractionRangeType? extractionRangeType,
             string timezone,
             int? filterId,
             Granularity? granularity, 
-            int? transformId
+            int? transformId,
+            FillerKind fillerK,
```
Comment (AndreaCuneo) [inline Artesian/Artesian.SDK/Service/Query/QueryParamaters/ActualQueryParamaters.cs]: Please make the Filler similar to MostRecent, RangeSelection etc. So with a Type and a Config part.

File: Artesian/Artesian.SDK/Service/Query/QueryParamaters/MasQueryParamaters.cs [hunk 11]
```diff
@@ -27,20 +28,86 @@ public MasQueryParamaters()
         /// <param name="timezone"></param>
         /// <param name="filterId"></param>
         /// <param name="products"></param>
+        /// <param name="fillerK"></param>
```
Comment (AndreaCuneo) [inline Artesian/Artesian.SDK/Service/Query/QueryParamaters/MasQueryParamaters.cs]: group these into a MarketAssessmentValue class reusing an existing one (should be there).

File: Artesian/Artesian.SDK/Service/Query/QueryParamaters/QueryParamaters.cs [hunk 12]
```diff
@@ -62,6 +65,10 @@ public QueryParamaters()
         /// filterId
         /// </summary>
         public int? FilterId { get; set; }
+        /// <summary>
+        /// Filler Kind
+        /// </summary>
+        public FillerKind FillerKind { get; set; } = new FillerKind();
```
Comment (AndreaCuneo) [inline Artesian/Artesian.SDK/Service/Query/QueryParamaters/QueryParamaters.cs]: FillerKind isn't an Enum? better to write FillerKind.Default then using `new`, as is clearer to read

File: Artesian/Artesian.SDK/Service/Query/IVersionedQuery.cs (hunk 9 above)
Comment (MichaelGeraghty) [reply Artesian/Artesian.SDK/Service/Query/IVersionedQuery.cs]: Method defaults to null "public VersionedQuery ForMUV(LocalDateTime? versionLimit = null)" allowing for .ForMUV()  use. Will the interface need to be defaulted to null also?

File: Artesian/Artesian.SDK/Service/Query/QueryParamaters/ActualQueryParamaters.cs (hunk 10 above)
Comment (MichaelGeraghty) [reply Artesian/Artesian.SDK/Service/Query/QueryParamaters/ActualQueryParamaters.cs]: Granularity? granularity, 
int? transformId,
FillerKindType fillerK,
FillerConfig fillerConfig

config class added with default value and period but due to naming ambiguity 
will have to distinguish between  public double? FillerTimeSeriesDV { get; set; } and
public MarketAssessmentValue FillerMasDV { get; set; }. Not sure of this.

File: Artesian/Artesian.SDK/Service/Query/IVersionedQuery.cs (hunk 9 above)
Comment (AndreaCuneo) [reply Artesian/Artesian.SDK/Service/Query/IVersionedQuery.cs]: yes because if the user did

```
IActualQuery q = qs.CreateActual();
q.ForMUV();
```

would break. And since that was possible, it must be preserved.

Comment (MichaelGeraghty) [reply Artesian/Artesian.SDK/Service/Query/IVersionedQuery.cs]: Ok , thanks for the clarification.

File: Artesian/Artesian.SDK/Service/Query/QueryParamaters/ActualQueryParamaters.cs (hunk 10 above)
Comment (AndreaCuneo) [reply Artesian/Artesian.SDK/Service/Query/QueryParamaters/ActualQueryParamaters.cs]: It does make sense. Ok with that duplication, including the name of the Fields
